app/src/main/java/io/appium/uiautomator2/handler/ListWindows.java

```diff
@@ -0,0 +1,271 @@
+                    Long physicalDisplayId = getPhysicalDisplayId(displayManager, displayId);
+                    return windowsOnAllDisplays.get(displayId).stream()
+                            .map(window -> createWindowModel(window, displayId, physicalDisplayId, skipScreenshots));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private List<WindowModel> listWindowsOnApiLevelBelow30(@Nullable DisplayManager displayManager, boolean skipScreenshots) {
+        List<AccessibilityWindowInfo> windowList = CustomUiDevice
+                .getInstance()
+                .getUiAutomation()
+                .getWindows();
+
+        // For API < 30, getWindows() only returns windows from the default display
+        int displayId = Display.DEFAULT_DISPLAY;
+        Long physicalDisplayId = getPhysicalDisplayId(displayManager, displayId);
+
+        return windowList.stream()
+                .map(window -> createWindowModel(window, displayId, physicalDisplayId, skipScreenshots))
+                .collect(Collectors.toList());
+    }
+
+    @Nullable
+    private DisplayManager getDisplayManager() {
+        return (DisplayManager) getInstrumentation().getTargetContext()
+                .getSystemService(Service.DISPLAY_SERVICE);
+    }
+
+    @Nullable
+    private Long getPhysicalDisplayId(@Nullable DisplayManager displayManager, int displayId) {
+        if (displayManager == null) {
+            return null;
+        }
+        Display display = displayManager.getDisplay(displayId);
+        return display != null ? DisplayIdHelper.getPhysicalDisplayId(display) : null;
+    }
+
+    private WindowModel createWindowModel(
+        AccessibilityWindowInfo window,
+        int displayId,
+        @Nullable Long physicalDisplayId,
+        boolean skipScreenshots
+    ) {
+        Integer windowId = window.getId();
+        Rect bounds = new Rect();
+        window.getBoundsInScreen(bounds);
+
+        String packageName = null;
+        AccessibilityNodeInfo root = window.getRoot();
+        if (root != null) {
+            packageName = StringHelpers.charSequenceToNullableString(root.getPackageName());
+        }
+
+        String screenshot = skipScreenshots ? null : takeWindowScreenshot(window);
+
+        return new WindowModel(windowId, displayId, physicalDisplayId, bounds, packageName, screenshot);
+    }
+
+    @Nullable
+    private String takeWindowScreenshot(AccessibilityWindowInfo windowInfo) {
+        // Screenshot API is only available from API 34
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            return null;
+        }
+
+        try {
+            Window window = getWindowFromAccessibilityWindowInfo(windowInfo);
+            if (window == null) {
+                return null;
+            }
+
+            UiAutomation uiAutomation = CustomUiDevice.getInstance().getUiAutomation();
+            Bitmap screenshot = uiAutomation.takeScreenshot(window);
```

Comment:
I see... They could have null cases depending on the condition

https://developer.android.com/reference/android/app/UiAutomation#takeScreenshot(android.view.Window)

> Used to capture a screenshot of a Window. This can return null in the following cases: 1. Window content hasn't been layed out. 2. Window doesn't have a valid SurfaceControl 3. An error occurred in SurfaceFlinger when trying to take the screenshot.

We could address this in the command's readme or log for `null` case (for future user support)

Reply:
yep, will mention it in the method readme